### ISSUE TITLE
Generate two reduction loops for vectorization

### ISSUE THREAD

**Issue by shunting314** [pytorch/pytorch#148402]
### 🐛 Describe the bug

A general reduction is to reduce each row of a [xnumel, rnumel] 2D tensors (multiple dimensional cases can be just treated as 2D tensor by flattening reduction and non-reductoin dimensions). When rnumel is not well aligned (128 bytes aligned), Inductor will pad the strides of the tensor to make the memory access more efficient.

E.g. if rnumel=50257, for bf16 tensor, Inductor pad the strides to the next multiple of 64 elements and we get 50304. The tensor's shape is not changed. Only strides get padded.

There are 2 problems if we do reduction for such a tensor:
1. There is a triton bug that such reduction will have un-coalesced memory access due to the non-const mask for a potential vectorized load.  ( https://github.com/pytorch/pytorch/issues/122840 )
2. The load is not vectorized and can be less efficient

Here is an optimization to fix it. We can split the reduction loop to 2 loops. The first loop is the main loop. It iterates all the elements up-to `rnumel_rounded = rnumel // RBLOCK * RBLOCK`. The second loop handles the left over elements not handled by the main loop.

Example code:
```
def triton_red_fused__log_softmax_16_manually_modified_to_two_loops(in_out_ptr0, in_ptr0, out_ptr0, xnumel, r0_numel, XBLOCK : tl.constexpr, R0_BLOCK : tl.constexpr):
    xnumel = 32768
    r0_numel = 50257
    rnumel = r0_numel
    RBLOCK: tl.constexpr = R0_BLOCK
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:, None]
    xmask = tl.full([XBLOCK, R0_BLOCK], True, tl.int1)
    r0_base = tl.arange(0, R0_BLOCK)[None, :]
    rbase = r0_base
    x0 = xindex
    _tmp3_max = tl.full([XBLOCK, R0_BLOCK], float('-inf'), tl.float32)
    _tmp3_sum = tl.zeros([XBLOCK, R0_BLOCK], tl.float32)

    # first loop
    r0_numel_round = (rnumel // R0_BLOCK) * R0_BLOCK
    for r0_offset in range(0, r0_numel_round, R0_BLOCK):
        r0_index = r0_offset + r0_base
        # r0_mask = r0_index < r0_numel
        roffset = r0_offset
        rindex = r0_index
        r0_1 = r0_index
        tmp0 = tl.load(in_ptr0 + (r0_1 + 50304*x0), None, eviction_policy='evict_first').to(tl.float32)
        tmp1 = tmp0.to(tl.float32)
        tmp2 = tl.broadcast_to(tmp1, [XBLOCK, R0_BLOCK])

        _tmp3_max_next, _tmp3_sum_next = triton_helpers.online_softmax_combine(
            _tmp3_max, _tmp3_sum, tmp2, True
        )

        _tmp3_max = _tmp3_max_next
        _tmp3_sum = _tmp3_sum_next

    # second loop
    for r0_offset in range(r0_numel_round, r0_numel, R0_BLOCK):
        r0_index = r0_offset + r0_base
        r0_mask = r0_index < r0_numel
        roffset = r0_offset
        rindex = r0_index
        r0_1 = r0_index
        tmp0 = tl.load(in_ptr0 + (r0_1 + 50304*x0), r0_mask, eviction_policy='evict_first', other=0.0).to(tl.float32)
        tmp1 = tmp0.to(tl.float32)
        tmp2 = tl.broadcast_to(tmp1, [XBLOCK, R0_BLOCK])

        _tmp3_max_next, _tmp3_sum_next = triton_helpers.online_softmax_combine(
            _tmp3_max, _tmp3_sum, tmp2, True
        )

        _tmp3_max = tl.where(r0_mask, _tmp3_max_next, _tmp3_max)
        _tmp3_sum = tl.where(r0_mask, _tmp3_sum_next, _tmp3_sum)


    tmp5, tmp6 = triton_helpers.online_softmax_reduce(
        _tmp3_max, _tmp3_sum, 1, True)
    tmp5 = tmp5[:, None]
    tmp6 = tmp6[:, None]
    tmp3 = tmp5
    tmp4 = tmp6
    tl.store(out_ptr0 + (x0), tmp3, None)
    tmp7 = tl_math.log(tmp4)
    tl.debug_barrier()
    tl.store(in_out_ptr0 + (x0), tmp7, None)
```

This can guarantees that the first loop (which handles majority of elements) to be fully vectorized and having coalesced memory access.

A full example https://gist.github.com/shunting314/2fb1f5381b62b363d1046a2e05741e7b

Perf for softmax: 

Generate 1 loops: 3.013ms    3.294GB    1093.18GB/s
Generate 2 loops: 2.076ms    3.294GB    1586.94GB/s

There is about 1.5x speedup for this kernel.

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @aakhundov @jansel @eellison 

### Error logs

.

### Versions

.

cc @chauhang @penguinwu

**Comment by eellison** [comment]
CC @htyu, I know you mentioned loop peeling previously. I think the cleanest way to implement this would be with a loop peeling parameter to range.

**Comment by htyu** [comment]
This is a cool optimization! Yeah this can also be achieved with automatic loop peeling based on hints in Triton.

**Comment by shunting314** [comment]
Is there a plan to implement the mentioned optimization with loop peeling in triton? I think changing Inductor codegen to achieve this is convenient and not too bad.

**Comment by htyu** [comment]
We don't have a concrete plan to improve loop peeling atm. Modifying the inductor to achieve that sounds more realistic for now.

**Comment by plotfi** [comment]
@shunting314 @htyu I am interested to see the original Torch code that generated this. Do you have it?

I have another case I have been looking at for some time, but it seems it was hand modified at some point.

**Comment by shunting314** [comment]
@plotfi  here it is: https://github.com/karpathy/llm.c/blob/master/train_gpt2.py#L180

**Comment by shunting314** [comment]
karthickai shows interest on this one. Assign to myself for now since I don't find karthickai on github.

**Comment by plotfi** [comment]
@shunting314 What would the triton_red_fused__log_softmax_16_manually_modified_to_two_loops example looked like without splitting into two loops? I mean, if it were all one reduction?

**Comment by shunting314** [comment]
@plotfi the full example link in the issue summary contains both kernels.

**Comment by liqiangxl** [comment]
Hi @shunting314, 
The 20% codegen overhead is removed by replacing the two-pass clone approach with a `single-pass masked→unmasked line map` (same pattern as xmask_unswitch in PR #178318). 

During the normal masked codegen pass, `load()`, `store()`, and `reduction()` register `masked_line → unmasked_line` entries in `_unmasked_line_map`.  At loop emission, `_splice_unmasked()` uses `IndentedBuffer.map()` to substitute mapped lines for the main loop; the tail loop splices the originals via the existing `_emit_reduction_loops()` .

**Result:** 
1. codegen overhead eliminated (was +17%, now 0%).
2. Performance on bf16 softmax (32768, 50257) : 
(a) 1.42x speedup for static shape
(b) 1.22x speedup for dynamic shape (no vectorization due to lack of stride divisible by 16 hint)

See updated [design doc](https://github.com/liqiangxl/my_design_docs/blob/main/inductor/loop_peeling.md) for details.
